packages/vitest/src/node/pools/threads.ts

```diff
@@ -154,14 +155,27 @@ export function createThreadsPool(ctx: Vitest, { execArgv, env }: PoolProcessOpt
         const files = Object.values(filesByEnv).flat()
         const results: PromiseSettledResult<void>[] = []
 
-        if (isolated) {
-          results.push(...await Promise.allSettled(files.map(({ file, environment, project }) =>
+        const { isolatedFiles, nonIsolatedFiles } = files.reduce((acc, spec) => {
+          const isolatedProject = getConfig(spec.project).poolOptions?.threads?.isolate ?? isolated
+          const key = isolatedProject ? 'isolatedFiles' : 'nonIsolatedFiles'
+          acc[key].push(spec)
+          return acc
+        }, { isolatedFiles: [] as GroupedSpec[], nonIsolatedFiles: [] as GroupedSpec[] })
+
+        if (isolatedFiles.length) {
+          results.push(...await Promise.allSettled(isolatedFiles.map(({ file, environment, project }) =>
```

Comment:
It now looks the same as the one in `nonIsolatedFiles.length` 🤔 But it would actually break of the environment is different 🤔 

@AriPerkkio do we even need to check for isolation at pool level? 🤔

Reply:
Resolving `poolOptions` in workspaces is complicated. 😢  

What exactly is not working here? Setting `{ poolOptions: { threads: { isolate: false } } }` in workspace project? It should have been fixed in https://github.com/vitest-dev/vitest/pull/4765 already. I think we decided that project specific `isolate` doesn't affect how pool is run from the main thread. It only affects how worker handles isolation.

Reply:
The issue is not with resolving options. It's if we do resolve them, then I can see that we don't even need the `if(isolate)` because both branches look identical

Why don't we already group isolated files by environment?

Reply:
> Why don't we already group isolated files by environment?

I think that when isolation is enabled, there is no need to do any grouping when passing the files and options to pool. Each worker is always terminated after a test run so we can assume that environment is set up freshly. There should not be a need for (manual) `pool.recycleWorkers()` calls when isolation is enabled. 

When isolation is disabled, we run as many files in same worker as possible. If non-isolated files have different enviroments, we'll still need to recycle workers between those groups.